README
Bio-RNA-BarMap

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Bio::RNA::BarMap

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        https://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-RNA-BarMap

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Bio-RNA-BarMap

    CPAN Ratings
        https://cpanratings.perl.org/d/Bio-RNA-BarMap

    Search CPAN
        https://metacpan.org/release/Bio-RNA-BarMap


LICENSE AND COPYRIGHT

Copyright (C) 2019 Felix Kuehnl

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.



Changes
Revision history for Bio-RNA-BarMap

0.03    2022-12-08
        Added GPLv3 LICENSE file.
0.02    2022-01-13
        - fixed "multiplication with hash" errors encountered with old Perl
          versions in two tests
        - increased parsing speed of mappings by about 20%

0.01    2021-12-22
        First version, released on an unsuspecting world.



Build.PL
use 5.012;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Bio::RNA::BarMap',
    license             => 'gpl_3',
    dist_author         => q{Felix Kuehnl <felix@bioinf.uni-leipzig.de>},
    dist_version_from   => 'lib/Bio/RNA/BarMap.pm',
    release_status      => 'stable',
    configure_requires => {
        'Module::Build' => '0',
    },
    build_requires => {
        'Test::More'      => '0',
        'Test::Exception' => '0.43',
    },
    requires => {
        'Moose'                     => 2.2011,
        'MooseX::StrictConstructor' => 0.21,
        'namespace::autoclean'      => 0.28,
        'autodie'                   => 2.29,
        'Scalar::Util'              => 1.50,
        'List::Util'                => 1.50,
        'File::Spec'                => 3.75,
        'Set::Scalar'               => 1.29,
        'IPC::System::Simple'       => 1.25,
    },
    add_to_cleanup     => [ 'Bio-RNA-BarMap-*' ],
);

$builder->create_build_script();


lib/Bio/RNA/BarMap.pm
package Bio::RNA::BarMap;
our $VERSION = '0.03';

use 5.012;
use warnings;

use Bio::RNA::BarMap::Mapping;
use Bio::RNA::BarMap::Mapping::Type;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Bio::RNA::BarMap - Parse and query I<BarMap> mappings.

=head1 SYNOPSIS

    use v5.12;              # for 'say()' and '//' a.k.a. logical defined-or
    use Bio::RNA::BarMap;

    # Parse a BarMap output file.
    my $barmap_file = 'N1M7_barmap_1.out';     # e.g. the test data in t/data/
    my $mapping = Bio::RNA::BarMap::Mapping->new($barmap_file);

    # Print mapped Barriers files.
    say join q{ }, $mapping->mapped_files;

    # Map a file.
    my $from_file = '8.bar';            # one of the mapped .bar files
    my $to_file = $mapping->map_file($from_file);   # last file maps to undef
    say "$from_file is mapped to ", $to_file // "nothing";

    # Map minimum 2 from file 8.bar to the next file.
    my $from_min = 2;
    my $to_min = $mapping->map_min_step($from_file, $from_min);

    # Map to an arbitrary file (only forward direction!)
    $to_file = '15.bar';
    $to_min = $mapping->map_min($from_file, $from_min, $to_file);

    # Verify mapping type: is the mapping exact (->) or approximate (~>)?
    my ($type, $to_min2) = $mapping->map_min_step($from_file, $from_min);
    say "Min $from_min from file '$from_file' is mapped ",
         $type->is_exact  ? 'exactly' : 'approximately', " to min $to_min2";
    say "Mapping arrow: ", $type->arrow;

=head1 DESCRIPTION

This module provides auxiliary classes to parse, query and print the mapping
file generated by the RNA kinetics simulation tool I<BarMap>, developed
at the Institute of Theoretical Biochemistry (TBI) in Vienna.

Note that this module is B<not> developed and maintained by the authors of
I<BarMap>.

=head1 CLASSES & METHODS

Bio::RNA::BarMap provides several classes for working with I<BarMap>'s output.
The most relevant is L<Bio::RNA::BarMap::Mapping>, which is used to parse the
mapping file generated by I<BarMap>. The constructor C<new()> takes a path or
handle to a I<BarMap> file and creates a new mapping object, which is used to
perform all kinds of queries on the mapping file.

The mapping of a single state to the next I<Barriers> file can be either exact
or approximate, as is indicated by a straight (C<-E<gt>>) or curved
(C<~E<gt>>) arrow in the mapping file. To provide this information, the
C<map_min()> and C<map_min_step()> methods of the mapping object -- when
called in a list context -- return not only the target minimum of the mapping,
but also an object of type L<Bio::RNA::BarMap::Mapping::Type>. The type object
can be queried using its methods C<is_exact()> and C<is_approx()>, and also
converted back to its arrow representation using C<arrow()>.

For more information, please refer to the documentation of the individual
classes.


=head1 AUTHOR

Felix Kuehnl, C<< <felix at bioinf.uni-leipzig.de> >>


=head1 BUGS

Please report any bugs or feature requests by raising an issue at
L<https://github.com/xileF1337/Bio-RNA-BarMap/issues>.

You can also do so by mailing to C<bug-bio-rna-barmap at rt.cpan.org>,
or through the web interface at
L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Bio-RNA-BarMap>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Bio::RNA::BarMap


You can also look for information at the official BarMap website:

L<https://www.tbi.univie.ac.at/RNA/bar_map/>


=over 4

=item * Github: the official repository

L<https://github.com/xileF1337/Bio-RNA-BarMap>

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-RNA-BarMap>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Bio-RNA-BarMap>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/Bio-RNA-BarMap>

=item * Search CPAN

L<https://metacpan.org/release/Bio-RNA-BarMap>

=back


=head1 LICENSE AND COPYRIGHT

Copyright 2019-2021 Felix Kuehnl.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.


=cut

# End of Bio/RNA/BarMap.pm



lib/Bio/RNA/BarMap/Mapping/Set.pm
# A simple set class implemented using a hash. Supports storing references.
# Faster then Set::Scalar for this specific use case. It significantly reduces
# the runtime.
package Bio::RNA::BarMap::Mapping::Set;
our $VERSION = '0.03';

use v5.12;
use warnings;

use autodie qw(:all);
use Moose;
use MooseX::StrictConstructor;
use namespace::autoclean;
use List::Util qw(pairmap);

# Elements are stored in a hash ref. For simple values, key is the element
# and value is undef. For references, the key gets stringified and the
# value stores the actual reference.
has _elems => (is => "ro", init_arg => undef, default => sub { {} });

# Return all elements. If defined, use the value, else the key.
sub elements { pairmap {$b // $a} %{ $_[0]->_elems } }

# Insert elements into the set. Returns itself.
sub insert {
    my $self = shift;
    # Don't store simple values twice, but preserve references.
    $self->_elems->{$_} = ref $_ ? $_ : undef foreach @_;
    $self;
}

__PACKAGE__->meta->make_immutable;

1; # End of Bio::RNA::BarMap::Mapping::Set


__END__

=pod

=encoding UTF-8

=head1 NAME

Bio::RNA::BarMap::Mapping::Set - Internally used class to store sets

=head1 SYNOPSIS

    use v5.12;              # for 'say()' and '//' a.k.a. logical defined-or
    use Bio::RNA::BarMap;

    # Construct new, empty set.
    my $set = Bio::RNA::BarMap::Mapping::Set->new();

    # Insert elements.
    $set->insert( qw(hello there foo) );

    # Retrieve elements.
    say "Elements in set: ", join q{, }, $set->elements;

=head1 DESCRIPTION

A simple, pure-Perl implementation of a set data structure. It is
significantly faster than L<Set::Scalar>, which became a bottleneck during the
implementation of this module. It supports storing references, but no deep
equality checks are performed.

=head1 METHODS

=head2 Bio::RNA::BarMap::Mapping::Set->new()

Returns a new, empty set. Adding elements during construction is currently not
supported (for no specific reason), use the C<insert()> method instead.

=head2 $set->elements()

Return all items contained in this set.

=head2 $set->insert(@elements)

Insert one or more C<@elements> into the set. Supports references.


=head1 AUTHOR

Felix Kuehnl, C<< <felix at bioinf.uni-leipzig.de> >>


=head1 BUGS

Please report any bugs or feature requests by raising an issue at
L<https://github.com/xileF1337/Bio-RNA-BarMap/issues>.

You can also do so by mailing to C<bug-bio-rna-barmap at rt.cpan.org>,
or through the web interface at
L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Bio-RNA-BarMap>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Bio::RNA::BarMap


You can also look for information at the official BarMap website:

L<https://www.tbi.univie.ac.at/RNA/bar_map/>


=over 4

=item * Github: the official repository

L<https://github.com/xileF1337/Bio-RNA-BarMap>

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-RNA-BarMap>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Bio-RNA-BarMap>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/Bio-RNA-BarMap>

=item * Search CPAN

L<https://metacpan.org/release/Bio-RNA-BarMap>

=back


=head1 LICENSE AND COPYRIGHT

Copyright 2019-2021 Felix Kuehnl.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.


=cut

# End of Bio/RNA/BarMap/Mapping/Set.pm


lib/Bio/RNA/BarMap/Mapping/Type.pm
# All methods are class methods, it's not meant to be instantiated. It just
# manages the type name string and converts it to arrows, etc.
package Bio::RNA::BarMap::Mapping::Type;
our $VERSION = '0.03';

use 5.012;
use warnings;

use Moose;
use Moose::Util::TypeConstraints;           # for enum()
use namespace::autoclean;


has '_type' => (
    is       => 'ro',
    isa      => enum([qw(EXACT APPROX)]),
    required => 1,
);

# Only allow construction from a single arrow string.
around BUILDARGS => sub {
    my $orig  = shift;
    my $class = shift;

    confess 'Pass a single arrow string to construct a Mapping::Type object'
        if @_ != 1 or ref $_[0];

    my $arrow = shift;                              # arrow string passed
    if ($arrow eq '->') {
        return $class->$orig(_type => 'EXACT')
    }
    elsif ($arrow eq '~>') {
        return $class->$orig(_type => 'APPROX')
    }
    else {
        confess 'Unknown arrow string in constructor';
    }
};

# Returns a new mapping type object of type 'exact'.
sub exact  { return $_[0]->new('->');  }

# Returns a new mapping type object of type 'approx'.
sub approx { return $_[0]->new('~>'); }

# Returns true iff the object is of EXACT type.
sub is_exact {
    my ($self) = @_;
    return $self->_type() eq 'EXACT';
}

# Returns true iff the object is of APPROX type.
sub is_approx {
    my ($self) = @_;
    return $self->_type() eq 'APPROX';
}

# Return the arrow representation of this object, i.e. '->' for exact and '~>'
# for approx mapping type objects.
sub arrow {
    my ($self) = @_;
    return $self->is_exact ? '->' : '~>';
}


__PACKAGE__->meta->make_immutable;

1;              # End of Bio::RNA::BarMap::Mapping::Type


__END__

=pod

=encoding UTF-8

=head1 NAME

Bio::RNA::BarMap::Mapping::Type - Represents the type of a I<BarMap> mapping
(exact or approximate)

=head1 SYNOPSIS

    use v5.12;                                              # for 'say'
    use Bio::RNA::BarMap;

    # Get a new mapping type object ... from arrow string:
    my $type = Bio::RNA::BarMap::Mapping::Type->new('->');  # exact
       $type = Bio::RNA::BarMap::Mapping::Type->new('~>');  # approx

    # ... or programmatically:
    $type = Bio::RNA::BarMap::Mapping::Type->exact;         # exact
    $type = Bio::RNA::BarMap::Mapping::Type->approx;        # approx

    # Verify mapping type.
    say 'Mapping is ', $type->is_exact  ? 'exact' : 'approximate';
    say 'Mapping arrow: ', $type->arrow;


=head1 DESCRIPTION

The objects of this class are used to represent the two possible types of a
mapping, either exact or approximate, and to easily convert between these two
values and their respective arrow representation. In I<BarMap> files,
C<< -> >> denotes an exact mapping, and C<< ~> >> an approximate one. The
object is constructed from the arrow string via the constructor C<new()>, or
using the methods C<exact()> and C<approx()> to get a new object of the
respective type.


=head1 METHODS

=head2 Bio::RNA::BarMap::Mapping::Type->new($arrow_string)

Constructs a new type object from its arrow string representation, i. e.
C<< -> >> for the exact and C<< ~> >> for the approximate variant.

=head2 Bio::RNA::BarMap::Mapping::Type->exact()

Class method. Returns a new mapping type object of I<exact> type.

=head2 Bio::RNA::BarMap::Mapping::Type->approx()

Class method. Returns a new mapping type object of I<approximate> type.

=head2 $type->is_exact()

Returns true iff the current object is of I<exact> type.

=head2 $type->is_approx()

Returns true iff the current object is of I<approximate> type.

=head2 $type->arrow()

Returns the type of this object in its arrow string representation, i. e.
C<< -> >> for the exact and C<< ~> >> for the approximate.

=head1 AUTHOR

Felix Kuehnl, C<< <felix at bioinf.uni-leipzig.de> >>

=head1 BUGS

Please report any bugs or feature requests by raising an issue at
L<https://github.com/xileF1337/Bio-RNA-BarMap/issues>.

You can also do so by mailing to C<bug-bio-rna-barmap at rt.cpan.org>,
or through the web interface at
L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Bio-RNA-BarMap>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Bio::RNA::BarMap


You can also look for information at the official BarMap website:

L<https://www.tbi.univie.ac.at/RNA/bar_map/>


=over 4

=item * Github: the official repository

L<https://github.com/xileF1337/Bio-RNA-BarMap>

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-RNA-BarMap>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Bio-RNA-BarMap>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/Bio-RNA-BarMap>

=item * Search CPAN

L<https://metacpan.org/release/Bio-RNA-BarMap>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2019 Felix Kuehnl.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.


=cut

# End of Bio::RNA::BarMap::Mapping::Type / lib/Bio/RNA/BarMap/Mapping/Type.pm


lib/Bio/RNA/BarMap/Mapping/FileMappingEntry.pm
# Mini class for entries of the file mapping hash.
package Bio::RNA::BarMap::Mapping::FileMappingEntry;
our $VERSION = '0.03';

use 5.012;
use warnings;

use Moose;
use namespace::autoclean;


has 'name' => (is => 'ro', required => 1);
# Ensure object is cleaned after use => use weak refs
has [qw(from to)] => (is => 'rw', weak_ref => 1);

__PACKAGE__->meta->make_immutable;

1;

__END__


=pod

=encoding UTF-8

=head1 NAME

Bio::RNA::BarMap::Mapping::FileMappingEntry - stores information about which
I<Barriers> file is mapped to which other I<Barriers> file.

=head1 SYNOPSIS

    use v5.12;                              # for 'say'
    use Bio::RNA::BarMap;

    my $file_mapping = Bio::RNA::BarMap::Mapping::FileMappingEntry->new(
        name => '1.bar',
        to   => '2.bar',
        from => undef,
    )

    say 'File ', $file_mapping->name, ' is mapped to ', $file_mapping->to;


=head1 DESCRIPTION

Internally used mini-class to store information about which I<Barriers> is
mapped to another. Stores both directions (mapped from, mapped to). The C<to>
and C<from> attributes may be undefined for the first and last file.

=head1 AUTHOR

Felix Kuehnl, C<< <felix at bioinf.uni-leipzig.de> >>


=head1 BUGS

Please report any bugs or feature requests by raising an issue at
L<https://github.com/xileF1337/Bio-RNA-BarMap/issues>.

You can also do so by mailing to C<bug-bio-rna-barmap at rt.cpan.org>,
or through the web interface at
L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Bio-RNA-BarMap>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Bio::RNA::BarMap


You can also look for information at the official BarMap website:

L<https://www.tbi.univie.ac.at/RNA/bar_map/>


=over 4

=item * Github: the official repository

L<https://github.com/xileF1337/Bio-RNA-BarMap>

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-RNA-BarMap>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Bio-RNA-BarMap>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/Bio-RNA-BarMap>

=item * Search CPAN

L<https://metacpan.org/release/Bio-RNA-BarMap>

=back


=head1 LICENSE AND COPYRIGHT

Copyright 2019-2021 Felix Kuehnl.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.


=cut

# End of Bio::RNA::BarMap::Mapping::FileMappingEntry


lib/Bio/RNA/BarMap/Mapping/MinMappingEntry.pm
# Bio/RNA/BarMap/Mapping/MinMappingEntry.pm

package Bio::RNA::BarMap::Mapping::MinMappingEntry;
our $VERSION = '0.03';

use 5.012;
use warnings;

use Moose;
use namespace::autoclean;

use Scalar::Util qw( weaken );

use Bio::RNA::BarMap::Mapping::Type;
use Bio::RNA::BarMap::Mapping::Set;


has 'index' => (is => 'ro', required => 1);

has 'to_type' => (
    is => 'rw',
    isa => 'Bio::RNA::BarMap::Mapping::Type',
);

# Ensure object is cleaned after use => use weak refs
has '_from'  => (
    is       => 'ro',
    init_arg => undef,      # use add_from() method to add elements
    default  => sub { Bio::RNA::BarMap::Mapping::Set->new },
);

has 'to' => (
    is          => 'rw',
    weak_ref    => 1,
    predicate   => 'has_to',
    isa         => __PACKAGE__,             # another entry
);

# Always use this method to add 'from' minima. This ensures the refs
# are weakened and no memory leaks arise.
sub add_from {
    my ($self, @from) = @_;
    weaken $_ foreach @from;                # turn into weak references
    $self->_from->insert(@from);
}

sub get_from {
    my ($self) = @_;
    my @from = $self->_from->elements;
    return @from;
}

__PACKAGE__->meta->make_immutable;

1;          # End of Bio::RNA::BarMap::Mapping::MinMappingEntry


__END__

=pod

=encoding UTF-8

=head1 NAME

Bio::RNA::BarMap::Mapping::MinMappingEntry - Store I<BarMap> mappings of a
single minimum.

=head1 SYNOPSIS

    use v5.12;              # for 'say()' and '//' a.k.a. logical defined-or
    use Bio::RNA::BarMap;

    my $entry = Bio::RNA::BarMap::Mapping::MinMappingEntry->new(
        index   => 3,               # of minimum of this entry
        to      => $to_min,         # the mininimum this one is mapped to
    );

    # Query the entry.
    if ($entry->has_to) {           # maps to something
        say 'This minimum maps ',
            $entry->$to_type->is_exact ? 'exactly' : 'approximately',
            ' to minimum ', $entry->to->index;
    }

    $entry->add_from($from_min_1, $from_min_2);     # add mins mapping to self
    say "Minima mapped to this minimum:",
    join q{, }, map {$_->index $entry->get_from();


=head1 DESCRIPTION

Internal class used to store the mapping of a single minimum. Both the forward
direction ("target minimum", C<to()>) and the reverse direction ("source
minima", C<get_from()>) are provided. While the target minimum is unique, but
not necessarily defined (cf. C<has_to()>), there may be zero to many source
minima, and so these are stored in a set internally. Use C<add_from()> to add
to this set.

=head1 METHODS

=head2 Bio::RNA::BarMap::Mapping::MinMappingEntry->new(arg_name => $arg_val, ...)

Constructor of the mapping entry class.

=over

=item Supported arguments:

=over

=item index

Required. Index of the minimum described by this entry.

=item to

Optional. Reference to mapping entry object describing the minimum that this
minimum is mapped to.

=back

=item Non-argument:

=over

=item from

To add source minima (i. e. minima that are mapped to this minimum), use the
method C<add_from()> instead.

=back

=back

=head2 $entry->index

Index of the minimum this entry is representing.

=head2 $entry->to_type

Type of the "to" mapping, either exact or approximate. Object of type
L<Bio::RNA::BarMap::Mapping::Type>.

=head2 $entry->to

Returns the entry this minimum is being mapped to. May be C<undef>.

=head2 $entry->to($to_min_entry)

Sets the C<to> attribute to point to C<$to_min_entry>.

=head2 $entry->add_from(@from_entries)

Adds entries to the set of source minima, i. e. those that are mapped to this
minimum. This method makes sure that the stored references are properly
weakened and no memory leaks arise.

=head2 $entry->get_from

Returns the entries of minima that are mapped to this minimum, as stored in
the source minima set.

=head1 AUTHOR

Felix Kuehnl, C<< <felix at bioinf.uni-leipzig.de> >>


=head1 BUGS

Please report any bugs or feature requests by raising an issue at
L<https://github.com/xileF1337/Bio-RNA-BarMap/issues>.

You can also do so by mailing to C<bug-bio-rna-barmap at rt.cpan.org>,
or through the web interface at
L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Bio-RNA-BarMap>.  I will be
notified, and then you'll automatically be notified of progress on your bug as
I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Bio::RNA::BarMap


You can also look for information at the official BarMap website:

L<https://www.tbi.univie.ac.at/RNA/bar_map/>


=over 4

=item * Github: the official repository

L<https://github.com/xileF1337/Bio-RNA-BarMap>

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=Bio-RNA-BarMap>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Bio-RNA-BarMap>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/Bio-RNA-BarMap>

=item * Search CPAN

L<https://metacpan.org/release/Bio-RNA-BarMap>

=back


=head1 LICENSE AND COPYRIGHT

Copyright 2019-2021 Felix Kuehnl.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see L<http://www.gnu.org/licenses/>.


=cut

# End of Bio/RNA/BarMap/Mapping/MinMappingEntry.pm
